### PR TITLE
Users

### PR DESCRIPTION
**Pull request by MalishkaMyu**
Again a lot of changes. Sorry!!!

### CODE REVIEW

File: .run/EmployeesdepApplication.run.xml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+<component name="ProjectRunConfigurationManager">
```
Comment (AlecKotovichSAM) [inline .run/EmployeesdepApplication.run.xml:1]: Remove this file from repository by updating .gitignore file

File: build.gradle [hunk 2]
```diff
@@ -18,6 +18,8 @@ dependencies {
 	implementation 'org.springframework.boot:spring-boot-starter-security'
 	implementation 'org.springframework.boot:spring-boot-starter-thymeleaf'
 	implementation 'org.springframework.boot:spring-boot-starter-web'
+	//Security
+	implementation 'org.springframework.security:spring-security-crypto:5.7.3'
```
Comment (AlecKotovichSAM) [inline build.gradle:23]: Move here implementation 'org.springframework.boot:spring-boot-starter-security' from above

File: build.gradle [hunk 3]
```diff
@@ -28,6 +30,7 @@ dependencies {
 	// SwaggerUI ( http://localhost:8080/swagger-ui/index.html )
 	implementation 'org.springdoc:springdoc-openapi-ui:1.6.11'
 
+
```
Comment (AlecKotovichSAM) [inline build.gradle]: Remove empty line

File: src/main/java/com/samsolutions/employeesdep/EmployeesdepApplication.java [hunk 4]
```diff
@@ -14,12 +18,30 @@ public static void main(String[] args) {
 		SpringApplication.run(EmployeesdepApplication.class, args);
 	}
 
+	@Value("${spring.flyway.placeholders.admin_login}")
+	private String adminLogin;
+
+	@Autowired
+    private MyPasswordEncoder encoder;
```
Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/EmployeesdepApplication.java:25]: Unformatted code

File: src/main/java/com/samsolutions/employeesdep/exception/GlobalExceptionHandler.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
+
+import java.util.Locale;
+
+@ControllerAdvice
+public class GlobalExceptionHandler extends ResponseEntityExceptionHandler {
+    @ExceptionHandler(EmployeesdepGlobalException.class)
+    @SuppressWarnings("rawtypes")
```
Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/exception/GlobalExceptionHandler.java]: Change ResponseEntity  to appropriate ResponseEntity<WHICH TYPE?>

File: src/main/java/com/samsolutions/employeesdep/exception/GlobalExceptionHandler.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+        return ResponseEntity
+                .badRequest()
+                .body(ErrorResponse.builder()
+                        .code(employeesdepGlobalException.getCode())
+                        .message(employeesdepGlobalException.getMessage())
+                        .build());
+    }
+
+    @ExceptionHandler({Exception.class})
+    @SuppressWarnings("rawtypes")
```
Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/exception/GlobalExceptionHandler.java]: same

File: src/main/java/com/samsolutions/employeesdep/model/converters/UserDTOToEntityConverter.java [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+import com.samsolutions.employeesdep.model.entities.User;
+import org.springframework.beans.BeanUtils;
+import org.springframework.core.convert.converter.Converter;
+
+public class UserDTOToEntityConverter implements Converter<UserDTO, User> {
+    @Override
+    public User convert(UserDTO source) {
+        User target = new User();
+        BeanUtils.copyProperties(source, target,"passwordHash");
+        if (!source.getPasswordHash().isBlank())
```
Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/model/converters/UserDTOToEntityConverter.java]: if (!source.getPasswordHash().isBlank())
            target.setPasswordHash(source.getPasswordHash());

->
 if (!source.getPasswordHash().isBlank()) {
            target.setPasswordHash(source.getPasswordHash());
}

File: src/main/java/com/samsolutions/employeesdep/model/converters/UserEntityToDTOConverter.java [hunk 8]
```diff
@@ -0,0 +1,16 @@
+package com.samsolutions.employeesdep.model.converters;
+
+import com.samsolutions.employeesdep.model.dto.UserDTO;
+import com.samsolutions.employeesdep.model.entities.User;
+import org.springframework.beans.BeanUtils;
+import org.springframework.core.convert.converter.Converter;
+
+public class UserEntityToDTOConverter implements Converter<User, UserDTO> {
+    @Override
```
Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/model/converters/UserEntityToDTOConverter.java:9]: Add empty line

File: src/main/java/com/samsolutions/employeesdep/model/services/UserService.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
+package com.samsolutions.employeesdep.model.services;
+
+import com.samsolutions.employeesdep.model.dto.UserDTO;
+
+import java.util.List;
+
+public interface UserService {
+    UserDTO createUser(UserDTO userToSaveDTO);
```
Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/model/services/UserService.java:9]: Add empty line

File: src/main/java/com/samsolutions/employeesdep/model/services/impl/UserServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+                    "The email " + userToSave.getEmail() + " is already registered for another user. " +
+                            "Please choose another email.");
+        }
+
+        // encoding password
+        userToSave.setPasswordHash(encoder.encode(userToSaveDTO.getPassword()));
+        // saving user
+        userRepository.save(userToSave);
+        UserDTO savedUserDTO = new UserEntityToDTOConverter().convert(userToSave);
+        if (savedUserDTO != null)
```
Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/model/services/impl/UserServiceImpl.java]: {}

File: src/main/resources/dbscripts/V1.4__new_users_table.sql [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    login character varying(20) NOT NULL,
+    password character varying(100) NOT NULL,
+    email character varying(70) NOT NULL,
+    created_at timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
+    updated_at timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
+    CONSTRAINT "users_pkey" PRIMARY KEY (user_id),
+    CONSTRAINT "users_unique_login" UNIQUE (login),
+    CONSTRAINT "users_unique_email" UNIQUE (email)
+);
+ALTER TABLE IF EXISTS "employees"
```
Comment (AlecKotovichSAM) [inline src/main/resources/dbscripts/V1.4__new_users_table.sql:17]: Write comment why this is needed

File: src/test/java/com/samsolutions/employeesdep/TestDataHelper.java [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+import org.springframework.stereotype.Component;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+
+@Component
+public class TestDataHelper {
+    @Autowired
```
Comment (AlecKotovichSAM) [inline src/test/java/com/samsolutions/employeesdep/TestDataHelper.java:16]: Add empty line

File: src/test/java/com/samsolutions/employeesdep/repository/EmployeeRepositoryTest.java [hunk 13]
```diff
@@ -1,30 +1,37 @@
...
 import com.samsolutions.employeesdep.model.dao.JpaRoleDao;
 import com.samsolutions.employeesdep.model.entities.Department;
 import com.samsolutions.employeesdep.model.entities.Employee;
 import com.samsolutions.employeesdep.model.entities.Role;
+import com.samsolutions.employeesdep.model.entities.User;
 import com.samsolutions.employeesdep.model.enums.Gender;
 import com.samsolutions.employeesdep.model.repository.DepartmentRepository;
 import com.samsolutions.employeesdep.model.repository.EmployeeRepository;
+import com.samsolutions.employeesdep.model.repository.UserRepository;
+import org.junit.jupiter.api.*;
```
Comment (AlecKotovichSAM) [inline src/test/java/com/samsolutions/employeesdep/repository/EmployeeRepositoryTest.java]: *

File: src/test/java/com/samsolutions/employeesdep/repository/EmployeeRepositoryTest.java [hunk 14]
```diff
@@ -1,30 +1,37 @@
...
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.core.io.Resource;
+
+import java.io.IOException;
+import java.time.LocalDate;
+import java.util.List;
+import java.util.Set;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (AlecKotovichSAM) [inline src/test/java/com/samsolutions/employeesdep/repository/EmployeeRepositoryTest.java]: **

File: src/test/java/com/samsolutions/employeesdep/repository/EmployeeRepositoryTest.java [hunk 15]
```diff
@@ -163,32 +191,40 @@ public void testFindInDepartment() {
...
+            Set<Role> rolesToSave = Set.of(role);
+            empNew.setEmployeeRoles(rolesToSave);
+        }
+
+        // reading user and assign it to employee
+        if (userRepository.existsByLogin("nyusha")) {
+            User userNyusha = userRepository.findByLogin("nyusha").orElse(null);
+            empNew.setUser(userNyusha);
+        }
         // saving the new employee. There are 2 employees in the "Java Department" now
```
Comment (AlecKotovichSAM) [inline src/test/java/com/samsolutions/employeesdep/repository/EmployeeRepositoryTest.java]: Make unified

File: src/test/java/com/samsolutions/employeesdep/repository/UserRepositoryTest.java [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.data.domain.Sort;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (AlecKotovichSAM) [inline src/test/java/com/samsolutions/employeesdep/repository/UserRepositoryTest.java]: **

File: src/test/java/com/samsolutions/employeesdep/repository/UserRepositoryTest.java [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+            User admin = userRepository.findByLogin(adminLogin).orElse(null);
+            if (admin != null) {
+                assertEquals(adminLogin, admin.getLogin());
+                assertEquals(adminEmail, admin.getEmail());
+            }
+        }
+    }
+
+    @Test
+    void testExistByEmail() {
```
Comment (AlecKotovichSAM) [inline src/test/java/com/samsolutions/employeesdep/repository/UserRepositoryTest.java]: testExistsByEmail

File: src/test/java/com/samsolutions/employeesdep/services/UserServiceTest.java [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+            assertEquals("newnyusha@mail.ru", savedUser.getEmail());
+
+            // restore old values
+            userToSave.setPassword("nyushapwd");
+            userToSave.setEmail("nyusha@mail.ru");
+            userService.updateUser(userToSave);
+        }
+    }
+
+
```
Comment (AlecKotovichSAM) [inline src/test/java/com/samsolutions/employeesdep/services/UserServiceTest.java]: Remove duplicate empty line

**Review by AlecKotovichSAM — CHANGES_REQUESTED** [review]
Fix review

File: src/main/java/com/samsolutions/employeesdep/exception/GlobalExceptionHandler.java (hunk 5 above)
Comment (MalishkaMyu) [reply src/main/java/com/samsolutions/employeesdep/exception/GlobalExceptionHandler.java]: ResponseEntity\<Object\> is OK?

Comment (MalishkaMyu) [reply src/main/java/com/samsolutions/employeesdep/exception/GlobalExceptionHandler.java:13]: ResponseEntity<String>

File: .run/EmployeesdepApplication.run.xml (hunk 1 above)
Comment (MalishkaMyu) [reply .run/EmployeesdepApplication.run.xml:1]: I added a new String in .gitignore. Don't know wheather it's enougth

Comment (MalishkaMyu) [reply .run/EmployeesdepApplication.run.xml:1]: I haven't found, how to delete file from repository selbst, if it's already there

**Review by AlecKotovichSAM — APPROVED** [review]

**Pull request merged**